### ISSUE TITLE
Cross-device 'single file' copying on windows fails

### ISSUE THREAD

**Issue by haf-afa** [kubernetes-sigs/krew#357]
```
PS W:\> .\krew.exe install --manifest=krew.yaml
Installing plugin: krew
W1029 12:35:34.778118   28440 install.go:127] failed to install plugin "krew": install failed: failed while moving files to the installation directory: could not rename file from "C:\\Users\\h\\AppData\\Local\\Temp\\krew-temp-move316849743" to "W:\\.krew\\store\\krew\\v0.3.1": rename C:\Users\h\AppData\Local\Temp\krew-temp-move316849743 W:\.krew\store\krew\v0.3.1: The system cannot move the file to a different disk drive.
F1029 12:35:34.781121   28440 root.go:58] failed to install some plugins: [krew]
```

- Windows
- Following installation instructions
- Running the install command both from C: and W: fails

**Comment by corneliusweig** [comment]
So this is really odd.. golang produces a `*LinkError` on all platforms, which contains the errno from the move syscall. The error for moving files on Linux/Windows is a
```go
type Errno uintptr
```
and on both platforms, the errno for moving across devices is `EXDEV = 0x12` (see `zerrors_linux_amd64.go` and https://docs.microsoft.com/en-us/cpp/c-runtime-library/errno-doserrno-sys-errlist-and-sys-nerr?view=vs-2019).

So why does it not correctly determine that this is a failed move operation?

**Comment by corneliusweig** [comment]
@haf-afa Could you reproduce the error scenario and run with debug output enabled?
```sh
.\krew.exe install --manifest=krew.yaml --v 9
```
Maybe this gives a further hint what's going wrong here.

**Comment by andrewsav-datacom** [comment]
May or may not be related, I'm also cannot get it working on Windows following installation instructions but with a different output (this is a fresh install never used krew before):

```text
G:\krew>krew install --manifest=krew.yaml
You need to perform a migration to continue using krew.
Please run `kubectl krew system receipts-upgrade`
F1030 12:30:50.816179   16188 root.go:58] krew home outdated

G:\krew>kubectl krew system receipts-upgrade
Error: unknown command "krew" for "kubectl"
Run 'kubectl --help' for usage.
unknown command "krew" for "kubectl"
````

**Comment by ahmetb** [comment]
@andrewsav-datacom can you please open a new issue? It seems like you had an older version of krew.

**Comment by haf-afa** [comment]
```
PS W:\> .\krew.exe install --manifest=krew.yaml --v 9
I1030 13:48:20.113492   15356 root.go:98] detected windows, will check for old krew installations to clean up
I1030 13:48:20.114494   15356 root.go:111] could not find krew's own plugin receipt, skipping cleanup of stale krew installations
I1030 13:48:20.117497   15356 install.go:144] --manifest specified, not ensuring plugin index
I1030 13:48:20.136497   15356 install.go:113] Will install plugin: krew
Installing plugin: krew
I1030 13:48:20.143495   15356 install.go:59] Looking for installed versions
I1030 13:48:20.146494   15356 platform.go:43] Matching platform for labels(arch=amd64,os=windows)
I1030 13:48:20.157490   15356 platform.go:51] Found matching platform with index (3)
I1030 13:48:20.160495   15356 install.go:78] Install plugin krew at version=v0.3.1
I1030 13:48:20.172022   15356 install.go:96] Creating download staging directory "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-downloads\\krew"
I1030 13:48:20.186020   15356 downloader.go:36] Fetching "https://github.com/kubernetes-sigs/krew/releases/download/v0.3.1/krew.tar.gz"
I1030 13:48:21.099865   15356 downloader.go:43] Reading download data into memory
I1030 13:48:24.398028   15356 downloader.go:48] Read 17154648 bytes of download data into memory
I1030 13:48:24.398028   15356 verifier.go:52] Compare sha256 (8770d57ac67c33b8499bb95edc89399ae3e6d54eebc270da456cd7aa15562329) signed version
I1030 13:48:24.401029   15356 downloader.go:184] detected "application/x-gzip" file type
I1030 13:48:24.407030   15356 downloader.go:97] tar: extracting to "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-downloads\\krew"
I1030 13:48:24.408028   15356 downloader.go:115] tar: processing "./krew-darwin_amd64" (type=48, mode=-rwxrwxr-x)
I1030 13:48:24.408028   15356 downloader.go:130] tar: ensuring parent dirs exist for regular file, dir=C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew
I1030 13:48:24.514062   15356 downloader.go:147] tar: processed "./krew-darwin_amd64"
I1030 13:48:24.515085   15356 downloader.go:115] tar: processing "./krew-linux_amd64" (type=48, mode=-rwxrwxr-x)
I1030 13:48:24.517031   15356 downloader.go:130] tar: ensuring parent dirs exist for regular file, dir=C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew
I1030 13:48:24.632064   15356 downloader.go:147] tar: processed "./krew-linux_amd64"
I1030 13:48:24.633079   15356 downloader.go:115] tar: processing "./krew-linux_arm" (type=48, mode=-rwxrwxr-x)
I1030 13:48:24.642028   15356 downloader.go:130] tar: ensuring parent dirs exist for regular file, dir=C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew
I1030 13:48:24.749288   15356 downloader.go:147] tar: processed "./krew-linux_arm"
I1030 13:48:24.749288   15356 downloader.go:115] tar: processing "./krew-windows_amd64.exe" (type=48, mode=-rwxrwxr-x)
I1030 13:48:24.752058   15356 downloader.go:130] tar: ensuring parent dirs exist for regular file, dir=C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew
I1030 13:48:24.944028   15356 downloader.go:147] tar: processed "./krew-windows_amd64.exe"
I1030 13:48:24.945055   15356 downloader.go:149] tar extraction to C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew complete
I1030 13:48:24.947065   15356 move.go:156] Creating plugin installation directory "W:\\.krew\\store\\krew\\v0.3.1"
I1030 13:48:24.961044   15356 move.go:162] Creating temp plugin move operations dir "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611"
I1030 13:48:24.961044   15356 move.go:125] Finding move targets from "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-downloads\\krew" to "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611" with file operation=index.FileOperation{From:"./krew-windows_amd64.exe", To:"krew.exe"}
I1030 13:48:24.963048   15356 move.go:44] Trying to move single file directly from="C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-downloads\\krew" to="C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611" with file operation=index.FileOperation{From:"./krew-windows_amd64.exe", To:"krew.exe"}
I1030 13:48:24.964089   15356 move.go:48] Detected single move from file operation=index.FileOperation{From:"./krew-windows_amd64.exe", To:"krew.exe"}
I1030 13:48:24.964089   15356 move.go:132] Move file from "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-downloads\\krew\\krew-windows_amd64.exe" to "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611\\krew.exe"
I1030 13:48:24.966027   15356 move.go:141] Move operations are complete
I1030 13:48:24.970025   15356 move.go:172] Move directory "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611" to "W:\\.krew\\store\\krew\\v0.3.1"
I1030 13:48:24.971025   15356 move.go:192] There's already a directory at move target "W:\\.krew\\store\\krew\\v0.3.1". deleting.
I1030 13:48:24.981026   15356 move.go:196] Move target directory "W:\\.krew\\store\\krew\\v0.3.1" cleaned up
I1030 13:48:24.981026   15356 move.go:175] Cleaning up installation directory due to error during copying files
I1030 13:48:24.985027   15356 install.go:101] Deleting the download staging directory C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew
W1030 13:48:24.992036   15356 install.go:127] failed to install plugin "krew": install failed: failed while moving files to the installation directory: could not rename file from "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611" to "W:\\.krew\\store\\krew\\v0.3.1": rename C:\Users\hefe01\AppData\Local\Temp\krew-temp-move551465611 W:\.krew\store\krew\v0.3.1: The system cannot move the file to a different disk drive.
F1030 13:48:24.995040   15356 root.go:56] failed to install some plugins: [krew]
sigs.k8s.io/krew/cmd/krew/cmd.init.1.func1
        /home/travis/gopath/src/sigs.k8s.io/krew/cmd/krew/cmd/install.go:138
github.com/spf13/cobra.(*Command).execute
        /home/travis/gopath/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:762
github.com/spf13/cobra.(*Command).ExecuteC
        /home/travis/gopath/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:852
github.com/spf13/cobra.(*Command).Execute
        /home/travis/gopath/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:800
sigs.k8s.io/krew/cmd/krew/cmd.Execute
        /home/travis/gopath/src/sigs.k8s.io/krew/cmd/krew/cmd/root.go:54
main.main
        /home/travis/gopath/src/sigs.k8s.io/krew/cmd/krew/main.go:24
runtime.main
        /home/travis/.gimme/versions/go1.13.linux.amd64/src/runtime/proc.go:203
runtime.goexit
        /home/travis/.gimme/versions/go1.13.linux.amd64/src/runtime/asm_amd64.s:1357
```

**Comment by haf-afa** [comment]
```
PS W:\> .\krew.exe install --manifest=krew.yaml --v 9
I1030 13:48:20.113492   15356 root.go:98] detected windows, will check for old krew installations to clean up
I1030 13:48:20.114494   15356 root.go:111] could not find krew's own plugin receipt, skipping cleanup of stale krew installations
I1030 13:48:20.117497   15356 install.go:144] --manifest specified, not ensuring plugin index
I1030 13:48:20.136497   15356 install.go:113] Will install plugin: krew
Installing plugin: krew
I1030 13:48:20.143495   15356 install.go:59] Looking for installed versions
I1030 13:48:20.146494   15356 platform.go:43] Matching platform for labels(arch=amd64,os=windows)
I1030 13:48:20.157490   15356 platform.go:51] Found matching platform with index (3)
I1030 13:48:20.160495   15356 install.go:78] Install plugin krew at version=v0.3.1
I1030 13:48:20.172022   15356 install.go:96] Creating download staging directory "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-downloads\\krew"
I1030 13:48:20.186020   15356 downloader.go:36] Fetching "https://github.com/kubernetes-sigs/krew/releases/download/v0.3.1/krew.tar.gz"
I1030 13:48:21.099865   15356 downloader.go:43] Reading download data into memory
I1030 13:48:24.398028   15356 downloader.go:48] Read 17154648 bytes of download data into memory
I1030 13:48:24.398028   15356 verifier.go:52] Compare sha256 (8770d57ac67c33b8499bb95edc89399ae3e6d54eebc270da456cd7aa15562329) signed version
I1030 13:48:24.401029   15356 downloader.go:184] detected "application/x-gzip" file type
I1030 13:48:24.407030   15356 downloader.go:97] tar: extracting to "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-downloads\\krew"
I1030 13:48:24.408028   15356 downloader.go:115] tar: processing "./krew-darwin_amd64" (type=48, mode=-rwxrwxr-x)
I1030 13:48:24.408028   15356 downloader.go:130] tar: ensuring parent dirs exist for regular file, dir=C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew
I1030 13:48:24.514062   15356 downloader.go:147] tar: processed "./krew-darwin_amd64"
I1030 13:48:24.515085   15356 downloader.go:115] tar: processing "./krew-linux_amd64" (type=48, mode=-rwxrwxr-x)
I1030 13:48:24.517031   15356 downloader.go:130] tar: ensuring parent dirs exist for regular file, dir=C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew
I1030 13:48:24.632064   15356 downloader.go:147] tar: processed "./krew-linux_amd64"
I1030 13:48:24.633079   15356 downloader.go:115] tar: processing "./krew-linux_arm" (type=48, mode=-rwxrwxr-x)
I1030 13:48:24.642028   15356 downloader.go:130] tar: ensuring parent dirs exist for regular file, dir=C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew
I1030 13:48:24.749288   15356 downloader.go:147] tar: processed "./krew-linux_arm"
I1030 13:48:24.749288   15356 downloader.go:115] tar: processing "./krew-windows_amd64.exe" (type=48, mode=-rwxrwxr-x)
I1030 13:48:24.752058   15356 downloader.go:130] tar: ensuring parent dirs exist for regular file, dir=C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew
I1030 13:48:24.944028   15356 downloader.go:147] tar: processed "./krew-windows_amd64.exe"
I1030 13:48:24.945055   15356 downloader.go:149] tar extraction to C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew complete
I1030 13:48:24.947065   15356 move.go:156] Creating plugin installation directory "W:\\.krew\\store\\krew\\v0.3.1"
I1030 13:48:24.961044   15356 move.go:162] Creating temp plugin move operations dir "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611"
I1030 13:48:24.961044   15356 move.go:125] Finding move targets from "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-downloads\\krew" to "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611" with file operation=index.FileOperation{From:"./krew-windows_amd64.exe", To:"krew.exe"}
I1030 13:48:24.963048   15356 move.go:44] Trying to move single file directly from="C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-downloads\\krew" to="C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611" with file operation=index.FileOperation{From:"./krew-windows_amd64.exe", To:"krew.exe"}
I1030 13:48:24.964089   15356 move.go:48] Detected single move from file operation=index.FileOperation{From:"./krew-windows_amd64.exe", To:"krew.exe"}
I1030 13:48:24.964089   15356 move.go:132] Move file from "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-downloads\\krew\\krew-windows_amd64.exe" to "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611\\krew.exe"
I1030 13:48:24.966027   15356 move.go:141] Move operations are complete
I1030 13:48:24.970025   15356 move.go:172] Move directory "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611" to "W:\\.krew\\store\\krew\\v0.3.1"
I1030 13:48:24.971025   15356 move.go:192] There's already a directory at move target "W:\\.krew\\store\\krew\\v0.3.1". deleting.
I1030 13:48:24.981026   15356 move.go:196] Move target directory "W:\\.krew\\store\\krew\\v0.3.1" cleaned up
I1030 13:48:24.981026   15356 move.go:175] Cleaning up installation directory due to error during copying files
I1030 13:48:24.985027   15356 install.go:101] Deleting the download staging directory C:\Users\hefe01\AppData\Local\Temp\krew-downloads\krew
W1030 13:48:24.992036   15356 install.go:127] failed to install plugin "krew": install failed: failed while moving files to the installation directory: could not rename file from "C:\\Users\\hefe01\\AppData\\Local\\Temp\\krew-temp-move551465611" to "W:\\.krew\\store\\krew\\v0.3.1": rename C:\Users\hefe01\AppData\Local\Temp\krew-temp-move551465611 W:\.krew\store\krew\v0.3.1: The system cannot move the file to a different disk drive.
F1030 13:48:24.995040   15356 root.go:56] failed to install some plugins: [krew]
sigs.k8s.io/krew/cmd/krew/cmd.init.1.func1
        /home/travis/gopath/src/sigs.k8s.io/krew/cmd/krew/cmd/install.go:138
github.com/spf13/cobra.(*Command).execute
        /home/travis/gopath/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:762
github.com/spf13/cobra.(*Command).ExecuteC
        /home/travis/gopath/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:852
github.com/spf13/cobra.(*Command).Execute
        /home/travis/gopath/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:800
sigs.k8s.io/krew/cmd/krew/cmd.Execute
        /home/travis/gopath/src/sigs.k8s.io/krew/cmd/krew/cmd/root.go:54
main.main
        /home/travis/gopath/src/sigs.k8s.io/krew/cmd/krew/main.go:24
runtime.main
        /home/travis/.gimme/versions/go1.13.linux.amd64/src/runtime/proc.go:203
runtime.goexit
        /home/travis/.gimme/versions/go1.13.linux.amd64/src/runtime/asm_amd64.s:1357
```

**Comment by haf-afa** [comment]
@andrewsav-datacom @ahmetb I get that exact error after this error. I have to manually remove ~/.krew before I can reproduce the above error.

**Comment by ahmetb** [comment]
@corneliusweig the "single file move" code path (not glob move) seems to be the problem. It directly uses `os.Rename`. https://github.com/kubernetes-sigs/krew/blob/d67e08799abe4fe2b8d3210cddc8bb2277fe9e85/pkg/installation/move.go#L137

It appears like the handling of EXDRV is in [moveOrCopyDir](https://github.com/kubernetes-sigs/krew/blob/d67e08799abe4fe2b8d3210cddc8bb2277fe9e85/pkg/installation/move.go#L201). 

It appears like line 137 needs to use `moveOrCopyDir` as well?

**Comment by andrewsav-datacom** [comment]
I'll add some more information here. This is if I run the install command forgetting to escalate to "Admin":

```text
G:\krew>.\krew.exe install --manifest=krew.yaml
Installing plugin: krew
W1031 07:54:50.430335   11588 install.go:127] failed to install plugin "krew": install failed: failed to link installed plugin: failed to create a symlink form "C:\\Users\\asavinykh\\.krew\\bin" to "C:\\Users\\asavinykh\\.krew\\bin\\kubectl-krew.exe": symlink C:\Users\asavinykh\.krew\store\krew\v0.3.1\krew.exe C:\Users\asavinykh\.krew\bin\kubectl-krew.exe: A required privilege is not held by the client.
F1031 07:54:50.432357   11588 root.go:58] failed to install some plugins: [krew]
```

After that, escalating to admin gives the errors in my previous message.

After deleting `~/.krew` as haf-afa suggested the instillation finishes but then I'm getting:

```text
G:\krew>kubectl krew
not supported by windows
```
At this point I'd like to pass up on krew altogether and return in half a year, when those may be fixed. I hope that me reporting this experience along with commands I ran and response I got could help the community to nail it down and make it a bit more user-friendly. Many thanks!

**Comment by ahmetb** [comment]
What version of kubectl are you on?
Until very recently no plugins worked on windows due to a kubectl bug. Also we’d like to use your help and debug these issues, your feedback is welcomed.

**Comment by andrewsav-datacom** [comment]
I was quite sure that I was on recent kubectl but I was wrong. Docker on Windows installs it's own kubectl and put it high on the Path. I have resolved that in the past, but the recent Docker update seemed to re-apply these and I ended up using old kubectl as you suspected.

Once I fixed this again, `kubectl krew` started working. Thank you for your help.

**Comment by ahmetb** [comment]
@andrewsav-datacom did you also run into the cross-device installation issue?
Also did you need to run `kubectl krew system receipts-upgrade` to get it working?

**Comment by AndrewSav** [comment]
>@andrewsav-datacom did you also run into the cross-device installation issue?
No

>Also did you need to run kubectl krew system receipts-upgrade to get it working?
No

**Comment by AndrewSav** [comment]
>Until very recently no plugins worked on windows due to a kubectl bug

@ahmetb  Could you please specify minimum kubectl version requirement for Windows?

**Comment by ahmetb** [comment]
It seems v1.16.0.
https://github.com/kubernetes/kubernetes/pull/76227/commits/13be373ce70ee5b90da6989b84f1a6c507e40706

**Comment by ahmetb** [comment]
/retitle Cross-device 'single file' copying on windows fails

**Comment by ahmetb** [comment]
Some insights after quick investigation:

1. This happens because we use ioutil.TempDir which happens to point to c:\ and KREW_ROOT is elsewhere.

2. There's definitely a bug in how we handle 'single file' copies vs 'glob' copies wrt cross-device file copying.

3. One can easily use `subst K: W:\some\dir\.krew` to create a phony letter-drive in user land.

I'll try to spend more time to fix this and verify. Thanks for reporting again.

**Comment by AndrewSav** [comment]
>We could use a "KREW_ROOT/tmp" dir, but then we'd have to clean it up?

If I understand it correctly you will have to clean up no matter what, on windows files do not get cleaned up automatically even if they are in tempDir.

**Comment by ahmetb** [comment]
I think we already do cleanup temp dirs.

**Comment by maracaipe** [comment]
This issue also affects me. I had to change the `TMP` env var to the C: disk and it worked.

**Comment by ahmetb** [comment]
@maracaipe @AndrewSav we have a fix at #375!

Can you please try it? You can either build from PR branch or use this file by extracting it and using it to install krew: [krew-windows_amd64.exe.zip](https://github.com/kubernetes-sigs/krew/files/3833074/krew-windows_amd64.exe.zip)



Make sure to delete %USERPROFILE%\.krew directory first to clean things up after a failed installation (otherwise you may run into another issue #374).
